source/common/http/conn_manager_utility.cc

```diff
@@ -36,12 +36,17 @@ void ConnectionManagerUtility::mutateRequestHeaders(Http::HeaderMap& request_hea
                                                     Runtime::Loader& runtime,
                                                     const LocalInfo::LocalInfo& local_info) {
   // Clean proxy headers.
-  request_headers.removeConnection();
   request_headers.removeEnvoyInternalRequest();
   request_headers.removeKeepAlive();
   request_headers.removeProxyConnection();
   request_headers.removeTransferEncoding();
-  request_headers.removeUpgrade();
+
+  // If this is a WebSocket Upgrade request, do not remove the Connection and Upgrade headers,
+  // as we forward them verbatim to the upstream hosts.
+  if (!Utility::isWebSocketUpgradeRequest(request_headers)) {
```

Comment:
I thought about this more and I think we should only do this check if protocol is HTTP/1? It's not valid on h2 and we can avoid the perf hit of the extra header lookups and compares?